azurerm/resource_arm_hdinsight.go

```diff
@@ -0,0 +1,1151 @@
+	strcase "github.com/stoewer/go-strcase"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmHDInsightCluster() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmHDInsightClusterCreate,
+		Read:   resourceArmHDInsightClusterRead,
+		Update: resourceArmHDInsightClusterUpdate,
+		Delete: resourceArmHDInsightClusterDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"location": locationSchema(),
+
+			"resource_group_name": resourceGroupNameSchema(),
+
+			"cluster_version": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "~3.6",
+				ValidateFunc: validation.StringInSlice([]string{
+					"~3.6",
+					"3.6",
+					"3.5",
+					"3.4",
+					"3.3",
+					"3.2",
+					"3.1",
+					"3.0",
+				}, false),
+			},
+
+			//TODO: Add support for Windows
+			"os_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "Linux",
+				ValidateFunc: validation.StringInSlice([]string{
+					"Linux",
+				}, false),
+				DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+			},
+
+			"tier": {
+				Type:     schema.TypeString,
+				Required: true,
+				Default:  "Standard",
+				ValidateFunc: validation.StringInSlice([]string{
+					"Standard",
+				}, false),
+				DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+			},
+
+			//TODO: add support for more configurations i.e. core-site
+			"cluster_definition": {
+				Type:     schema.TypeList,
+				Required: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"blueprint": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"kind": {
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: validation.StringInSlice([]string{
+								"hadoop",
+								"spark",
+								"storm",
+								"hbase",
+								"rserver",
+								"kafka",
+								"interactivequery",
+								"interactivehive",
+							}, true),
+							DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+						},
+						"component_version": {
+							Type:     schema.TypeMap,
+							Computed: true,
+						},
+						"configurations": {
+							Type:     schema.TypeList,
+							Optional: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"gateway": {
+										Type:     schema.TypeList,
+										Optional: true,
+										Elem: &schema.Resource{
+											Schema: map[string]*schema.Schema{
+												"rest_auth_credential__is_enabled": {
+													Type:     schema.TypeBool,
+													Optional: true,
+												},
+												"rest_auth_credential__username": {
+													Type:     schema.TypeString,
+													Optional: true,
+												},
+												"rest_auth_credential__password": {
```

Comment:
This might be difficult - I need a programmatic way of mapping between a valid Terraform key i.e. [a-z_]+ and the dot separated key needed by the Azure resource. If I remove the '__' I have no way to detect where the '.' should be. Appreciate it's a hack but not sure on the best way around it.